src/EventListener/AdminEventListener.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\EventListener;
+
+use Sonata\AdminBundle\Admin\AdminInterface;
+use Sonata\AdminBundle\Admin\BreadcrumbsBuilderInterface;
+use Sonata\AdminBundle\Admin\Pool;
+use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\KernelEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Twig\Environment;
+
+/**
+ * @author Christian Gripp <mail@core23.de>
+ */
+final class AdminEventListener implements EventSubscriberInterface
+{
+    /**
+     * @var Environment
+     */
+    private $twig;
+
+    /**
+     * @var Pool
+     */
+    private $pool;
+
+    /**
+     * @var BreadcrumbsBuilderInterface
+     */
+    private $breadcrumbBuilder;
+
+    /**
+     * @var TemplateRegistryInterface
+     */
+    private $templateRegistry;
+
+    public function __construct(
+        Environment $twig,
+        Pool $pool,
+        TemplateRegistryInterface $templateRegistry,
+        BreadcrumbsBuilderInterface $breadcrumbBuilder
+    ) {
+        $this->twig = $twig;
+        $this->pool = $pool;
+        $this->templateRegistry = $templateRegistry;
+        $this->breadcrumbBuilder = $breadcrumbBuilder;
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            KernelEvents::REQUEST => [['onKernelRequest', 17]],
+        ];
+    }
+
+    public function onKernelRequest(KernelEvent $event): void
+    {
+        $request = $event->getRequest();
+
+        $admin = $this->getAdmin($request);
+
+        if (null === $admin) {
+            return;
+        }
+
+        if ($this->isXmlHttpRequest($request)) {
+            $baseTemplate = $this->templateRegistry->getTemplate('ajax');
+        } else {
+            $baseTemplate = $this->templateRegistry->getTemplate('layout');
+
+            $this->twig->addGlobal('breadcrumbs_builder', $this->breadcrumbBuilder);
+        }
+
+        $this->twig->addGlobal('base_template', $baseTemplate);
+        $this->twig->addGlobal('admin', $admin);
+        $this->twig->addGlobal('admin_pool', $this->pool);
+    }
+
+    /**
+     * @return AdminInterface<object>
+     *
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     */
+    private function getAdmin(Request $request): ?AdminInterface
```

Comment:
Can't we use the `adminFetcher` instead ?

Reply:
Done